[Refactor] Generically-typed StartNewAsync() overloads

brandonh-msft opened the pull request:
@cgillum Here's my attempt at a PR to alleviate the issue that others and I have ran in to while using the `StartNewAsync` call in Durable Functions.

Problem:
The following code:
```csharp
var instanceId = await client.StartNewAsync(@"MyOrchestrator", "thisInstanceId");
Debug.Assert(instanceId == @"thisInstanceId");
```

Will throw the assertion. This is because if a user wishes to specify the instanceId for a new orchestration they _must_ give some form of input - `null` or otherwise. While intellisense is the way one would realize what they're doing wrong, it's only after they attempt this and end up getting back an instanceId value other than the one they specified that they realize something's gone awry due to the fact that `string : object` so the `StartNewAsync(string, object)` overload gets hit and executed.

By using generics we can not only prevent this runtime confusion, but also enable an even simpler scenario: `StartNewAsync(@"MyOrchestrator")` which is an added benefit.

Hopefully you find this idea/PR valuable. Thanks! You'll notice I have the target for it as the `v2` branch due to its breaking nature.

cgillum commented:
Cool! One thing - can you clarify for me how the generic parameter helps? I'm not sure I understand what benefit it provides. I would assume that the example code you provided in your description would still have the same problem - i.e. the C# compiler infers `T` to be `string` in that case?

brandonh-msft commented:
no because if you tried to pass a `string` for one of the `input` values, you'd _have_ to do `<string>` on the method call or you'd get the "ambiguous reference" error

brandonh-msft commented:
d'oh you're right in its current state; I forgot an overload. Working on this now! 😊

brandonh-msft commented:
So to circle back. by not specifying the type parameter, and using the (string, string) overload (function name, instance id) you automatically go to a `null` type param and the instance id specified
if you wanted that to be (function, input) you'd now _have_ to specify `<string>` on the `StartNewAsync` call

cgillum commented on src/WebJobs.Extensions.DurableTask/ContextInterfaces/DurableContextExtensions.cs:
```diff
@@ -405,11 +406,49 @@ public static Task CallEntityAsync(this IDurableOrchestrationContext context, En
+        /// <exception cref="ArgumentException">
+        /// The specified function does not exist, is disabled, or is not an orchestrator function.
+        /// </exception>
+        public static Task<string> StartNewAsync(
```
nit: I would put this overload at the top of the list since it's the least specific.

cgillum commented on src/WebJobs.Extensions.DurableTask/ContextInterfaces/IDurableOrchestrationClient.cs:
```diff
@@ -117,12 +117,13 @@ public interface IDurableOrchestrationClient
         /// The specified function does not exist, is disabled, or is not an orchestrator function.
         /// </exception>
-        Task<string> StartNewAsync(string orchestratorFunctionName, string instanceId, object input);
+        Task<string> StartNewAsync<T>(string orchestratorFunctionName, string instanceId, T input);
```
Let's consider naming this `TInput` so that it's not confused to be used for the output (CallActivityAsync<T>'s T is for the output whereas this one is for the input).

ConnorMcMahon commented on src/WebJobs.Extensions.DurableTask/ContextInterfaces/DurableContextExtensions.cs:
```diff
@@ -396,7 +396,8 @@ public static Task CallEntityAsync(this IDurableOrchestrationContext context, En
         /// <param name="orchestratorFunctionName">The name of the orchestrator function to start.</param>
-        /// <param name="input">JSON-serializeable input value for the orchestrator function.</param>
+        /// <param name="instanceId">The ID to use for the new orchestration instance.</param>
+        /// <typeparam name="T">The type of the input value for the orchestrator function.</typeparam>
```
NIT: no type parameter in this method.

ConnorMcMahon commented on test/Common/DurableClientBaseTests.cs:
```diff
@@ -31,9 +28,28 @@ public async Task StartNewAsync_is_calling_overload_method()
+
+        [Fact]
+        [Trait("Category", PlatformSpecificHelpers.TestCategory)]
+        public async Task StartNewAsync_is_calling_overload_method_with_specific_instance_id()
```
NIT: if this test, and the other test in this file, could be renamed to match the naming scheme of our other tests, that would be great.

brandonh-msft replied:
can you elaborate? I kept the name for one, and added a new one that _seemed_ to follow the same convention....

![image](https://user-images.githubusercontent.com/20270743/67590966-1a582600-f711-11e9-93eb-ef47790cbdfb.png)

happy to make changes, though!

ConnorMcMahon replied:
I now realize we don't have a firm convention. I have a preference towards Capitalizing the first letter after each `_`, but that is not consistent across our tests, so I won't make you address that in this PR. 

Making it more consistent is for another PR :).

cgillum commented:
Connor's first "nit" isn't really a "nit" - it's causing the build to fail:

>  ContextInterfaces\DurableContextExtensions.cs(400,30): error CS1711: XML comment has a typeparam tag for 'T', but there is no type parameter by that name [C:\projects\azure-functions-durable-extension\src\WebJobs.Extensions.DurableTask\WebJobs.Extensions.DurableTask.csproj]

brandonh-msft commented:
Weird. Wonder why it was building on my box and in the previous CI? I'll have a look today and push the fix

ConnorMcMahon commented:
@brandonh-msft, our build is configured to treat warnings as errors when in Release build, which is what our CI uses. It's specifically designed to prevent people from checking in warnings into our main branches. You probably built as debug locally, which is why it built fine.

amdeel commented:
/AzurePipelines run

brandonh-msft commented:
@cgillum point of order on this. When I made these changes and did a build, a couple of XML files were also updated. Do you want those committed? Specifically:
```
modified:   src/WebJobs.Extensions.DurableTask/Microsoft.Azure.WebJobs.Extensions.DurableTask-net471.xml
modified:   src/WebJobs.Extensions.DurableTask/Microsoft.Azure.WebJobs.Extensions.DurableTask-netstandard2.0.xml
```

cgillum commented:
@brandonh-msft yes, those should also be committed.

cgillum commented:
/AzurePipelines run

brandonh-msft commented:
closed in favor of PR to `dev`
